src/common/number/numericString.ts

```diff
@@ -1,79 +1,155 @@
+        // Get the leftmost and largest match
+        if (
+          coreRadix < 0 ||
+          match.index < coreBegin ||
+          (match.index === coreBegin && match[0].length > coreLength)
+        ) {
+          coreBegin = match.index;
+          coreLength = match[0].length;
+          coreRadix = radix;
+          coreSign = match[1] === '-';
+        }
       }
+    }
 
-      // Regex to determine if this number has letters around it,
-      // if it doesn't then that is easy and no prefix or suffix is needed
-      let findNondigits = /[^\d-+]+/g;
-
-      // Regex to find any leading characters before the number
-      let findPrefix = /^[^\d-+]+(?=[0-9]+)/g;
-
-      // Regex to find any trailing characters after the number
-      let findSuffix = /[^\d]*[\d]*(.*)/g;
-      let newPrefix = prefix;
-      let newSuffix = '';
-      let newNum = input;
+    if (coreRadix < 0) {
+      return undefined;
+    }
 
-      // Only use this section if this is a number surrounded by letters
-      if (
-        findNondigits.exec(input) !== null &&
-        NumericString.matchings[NumericString.matchings.length - 1].regex === regex
-      ) {
-        let prefixFound = findPrefix.exec(input);
-        let suffixFound = findSuffix.exec(input);
+    const coreEnd = coreBegin + coreLength;
 
-        // Find the prefix if it exists
-        if (prefixFound !== null) {
-          newPrefix = prefixFound.toString();
-        }
+    const prefix = input.slice(0, coreBegin);
+    const core = input.slice(coreBegin, coreEnd);
+    const suffix = input.slice(coreEnd, input.length);
 
-        // Find the suffix if it exists
-        if (suffixFound !== null) {
-          newSuffix = suffixFound[1].toString();
-        }
+    let value = parseInt(core, coreRadix);
 
-        // Obtain just the number with no extra letters
-        newNum = newNum.slice(newPrefix.length, newNum.length - newSuffix.length);
-      }
+    // 0x00ff:  numLength = 4
+    // 077:     numLength = 2
+    // -0999:   numLength = 3
+    // The numLength is only useful for parsing non-decimal. Decimal with
+    // leading zero will be trimmed in `toString()`. If value is negative,
+    // remove the width of negative sign.
+    const numLength = coreLength - NumericString.numPrefix[coreRadix].length - (coreSign ? 1 : 0);
 
-      return new NumericString(parseInt(newNum, base), base, newPrefix, newSuffix);
+    // According to original vim's behavior, for hex and octal, the leading
+    // '-' *should* be captured and preseved but *should not* be regarded as
+    // part of number, which means with <C-a>, `-0xf` turns into `-0x10`. So
+    // for hex and octal, we make the value absolute and set the negative
+    // sign flag.
+    let negative = false;
+    if (coreRadix !== 10 && coreSign) {
+      value = -value;
+      negative = true;
     }
 
-    return undefined;
+    return [new NumericString(value, coreRadix, numLength, prefix, suffix, negative), coreEnd];
```

Comment:
Instead of returning a tuple, could you return an object, so it's clearer that the other value is the suffix offset?

Reply:
Done!